keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourcePersister.kt

```diff
@@ -60,32 +60,60 @@ class ResourcePersister(
         Environment(
           name = env.name,
           resources = env.resources.mapTo(mutableSetOf()) { resource ->
-            upsert(
-              resource.copy(
-                metadata = mapOf("serviceAccount" to deliveryConfig.serviceAccount) + resource.metadata
-              )
-            )
+            resource
+              .copy(metadata = mapOf("serviceAccount" to deliveryConfig.serviceAccount) + resource.metadata)
+              .normalize()
           },
           constraints = env.constraints,
           notifications = env.notifications
         )
       }
     )
+
+    validate(new)
+
+    new.resources.forEach { resource ->
+      upsert(resource)
+    }
     new.artifacts.forEach { artifact ->
       artifact.register()
     }
     deliveryConfigRepository.store(new)
+
     if (old != null) {
       cleaner.removeResources(old, new)
     }
     return new
   }
 
+  /**
+   * Validates that resources have unique ids, throws an exception if invalid
+   */
+  private fun validate(config: DeliveryConfig) {
+    val resources = config.environments.map { it.resources }.flatten().map { it.id }
+    val distinct = resources.distinct()
+
+    if (resources.size != distinct.size) {
+      val duplicates = resources.groupingBy { it }.eachCount().filter { it.value > 1 }.keys.toList()
+      val envToResources: Map<String, MutableList<String>> = config.environments
+        .map { env -> env.name to env.resources.map { it.id }.toMutableList() }.toMap()
+      val envsAndDuplicateResources = envToResources
+        .filterValues { rs: List<String> ->
+          // remove all the resources we don't care about from this mapping
+          rs.filter { it in duplicates }
```

Comment:
The result of this expression is not being assigned to anything, so my guess is that the line below doesn't have the intended effect...

Shouldn't the two statements be combined as follows?
```kotlin
rs.filter { it in duplicates }.isNotEmpty()
```

Reply:
🤦‍♂ yes!